pkg/klusterlet/clusterclaim/clusterclaimer.go

```diff
@@ -95,14 +96,64 @@ func init() {
 }
 
 type ClusterClaimer struct {
-	ClusterName    string
-	Product        string
-	Platform       string
-	HubClient      client.Client
-	KubeClient     kubernetes.Interface
-	ConfigV1Client openshiftclientset.Interface
-	OauthV1Client  openshiftoauthclientset.Interface
-	Mapper         meta.RESTMapper
+	ClusterName      string
+	Product          string
+	Platform         string
+	HubClient        client.Client
+	KubeClient       kubernetes.Interface
+	ConfigV1Client   openshiftclientset.Interface
+	OauthV1Client    openshiftoauthclientset.Interface
+	Mapper           meta.RESTMapper
+	managedclusterID string
+}
+
+func NewClusterClaimer(
+	clusterName string,
+	hubClient client.Client,
+	kubeClient kubernetes.Interface,
+	configV1Client openshiftclientset.Interface,
+	oauthV1Client openshiftoauthclientset.Interface,
+	mapper meta.RESTMapper,
+) (ClusterClaimer, error) {
+	managedClusterID, err := getManagedClusterID(kubeClient, mapper, configV1Client)
+	if err != nil {
+		return ClusterClaimer{}, err
+	}
+
+	return ClusterClaimer{
+		ClusterName:      clusterName,
+		HubClient:        hubClient,
+		KubeClient:       kubeClient,
+		ConfigV1Client:   configV1Client,
+		OauthV1Client:    oauthV1Client,
+		Mapper:           mapper,
+		managedclusterID: managedClusterID,
+	}, nil
+}
+
+// getManagedClusterID returns the managed cluster ID of the cluster.
+// We get ID by the sequence of: openshiftID, uid of kube-system namespace, random UID.
+func getManagedClusterID(kubeClient kubernetes.Interface, mapper meta.RESTMapper, configV1Client openshiftclientset.Interface) (string, error) {
```

Comment:
I could define like this :
```
func (c *ClusterClaimer) getManagedClusterID() (string, error) {
   if c.managedclusterID != "" {
       return c. managedclusterID
   }
   .....
}
```
it will not change the  `c.isOpenShift` method.

Reply:
Update, but that also means controller will send request to determine if this is an openshift every sync. Not sure whether we should take it into account.
Please take another look.